%%%%%%%%%%%%%%%%%%%%%%%%%
% PACKAGES              %
%%%%%%%%%%%%%%%%%%%%%%%%%
\documentclass[]{report} % book|article|…

\usepackage[utf8x]{inputenc}    % accents
\usepackage{geometry}           % marges
\usepackage[francais]{babel}    % langue
\usepackage{graphicx}           % images
\usepackage{verbatim}           % texte préformaté
\usepackage{fancyhdr}           % fancy
\usepackage{filecontents}       % write file directly
\usepackage{csvsimple}          % csv reader
\usepackage{lastpage}	        % get number of last page
\usepackage{listings}           % source code 
\usepackage{url}                % clickable urls 
\usepackage{float}              % exact placing of figures
\usepackage{amsmath}            % fancy matrices
\usepackage{xkeyval}            % keywords arguments for newcommand
\usepackage{xifthen}            % provides \isempty test

% packages for graphics
\usepackage{fancybox}         
\usepackage{pgfplots}        
\usepackage{pgfplotstable}
\usepgfplotslibrary{dateplot}
\usepackage{pgfplots}

\definecolor{Gene0}{RGB}{250, 164, 1}
\definecolor{Gene1}{RGB}{128, 0, 128}
\definecolor{Gene2}{RGB}{255, 0, 0}
\definecolor{Gene3}{RGB}{58, 242, 75}
\definecolor{Gene4}{RGB}{8, 81, 156}
\definecolor{Diversity}{RGB}{0, 0, 0}

\pgfplotscreateplotcyclelist{list}{
        {Gene0},
        {Gene1},
        {Gene2},
        {Gene3},
        {Gene4},
        {Diversity}
}




%%%%%%%%%%%%%%%%%%%%%%%%%
% SIMULATION COMMAND    %
%%%%%%%%%%%%%%%%%%%%%%%%%
%This is needed for encapsulating.
%changes the catcode of @ so it
%can be used freely (not needed
%when writing a package)
\makeatletter

\define@cmdkey [SIMPAR] {fam} {generations} [200] {}
\define@cmdkey [SIMPAR] {fam} {popsize}     [300] {}
\define@cmdkey [SIMPAR] {fam} {parents}     [2]   {}
\define@cmdkey [SIMPAR] {fam} {seed}        [4223]{}
\define@cmdkey [SIMPAR] {fam} {mutationrate}[0.0001] {}
\define@cmdkey [SIMPAR] {fam} {initialpheno}[1-11-11] {}
 
\presetkeys [SIMPAR] {fam} {generations = 1,
                            popsize = 2,
                            parents = 3,
                            seed = 4,
                            mutationrate = 5,
                            initialpheno = 6,
                           }{}

\newcommand{\simulationParameters}[1]{%
        \setkeys[SIMPAR]{fam}{#1}
        \begin{tabular}{|l|c||l|c|}
        \hline
        \textbf{Generations computed} & \cmdSIMPAR@fam@generations & \textbf{Population size}     & \cmdSIMPAR@fam@popsize \\ \hline
        \textbf{Parent number}        & \cmdSIMPAR@fam@parents     & \textbf{Seed}                & \cmdSIMPAR@fam@seed    \\ \hline
        \textbf{Mutation rate}        & \cmdSIMPAR@fam@mutationrate& \textbf{Initial Phenotype}   & \cmdSIMPAR@fam@initialpheno \\ \hline
        \end{tabular}
}

%Change again the @ catcode
%to normal
\makeatother




%%%%%%%%%%%%%%%%%%%%%%%%%
% PRÉAMBULE             %
%%%%%%%%%%%%%%%%%%%%%%%%%
\title{Genomat\\PRJ project Report}
\author{Ostiane D'AUGUSTIN and Lucas BOURNEUF}
% laisser vide pour date de compilation
\date{} 

% FORMAT PAGES         
\pagestyle{fancy} % nom du rendu (définit les lignes suivantes)
        \lhead{M1 BIG} % left head
        \chead{} % center head
        \rhead{2014/2015} % right head
        \lfoot{} % left foot
        \cfoot{\thepage/\pageref{LastPage}} % center foot
        \rfoot{} % right foot






%%%%%%%%%%%%%%%%%%%%%%%%%
% BEGIN                 %
%%%%%%%%%%%%%%%%%%%%%%%%%
\begin{document}
        \maketitle % page de titre




%%%%%%%%%%%%%%%%%%%%%%%%%
% SECTION               %
%%%%%%%%%%%%%%%%%%%%%%%%%
\section*{Introduction}
	\paragraph*{}
	The aim of this project is to implement a genetic algorithm which will bring out the connexions between different genes, 
        their regulation networks, and some properties linked to, as well as their resistance to genes knock-out (KO).
	\paragraph*{}
	We will study a famous genes network example, which has been highlighted by Wagner. Wagner's genes network is an artificial genes network calculation model. 
        It has explained all the development and evolution process of genetic regulation networks. It has been first developed by Andreas Wagner in 1996, 
        and then used by several research teams in order to study genes network evolution, genes expression, etc.
	\paragraph*{}
	We will simulate \textit{in silico} a population's evolution over many generations. Each individual is caracterized by his genotype 
        (that is to say, its genes network). 
        Individuals will be submit to an evolutive process by the intermediary of their genes networks.




%%%%%%%%%%%%%%%%%%%%%%%%%
% SECTION               %
%%%%%%%%%%%%%%%%%%%%%%%%%
\section*{Project organization}
	\paragraph*{}
        The project follows a Python-like packaging, with the main program launchable as a package. (\textit{python -m genomat})
        Many parameters can be sended through command-line arguments, or configuration file.
        Outputs are mainly in \textsc{csv format}. (notabily those provided for statistics)

	\paragraph*{}
        A second and incomplete implementation of Genomat is available under the name of \textit{genomat\_func}. 
        This implementation is only here as a proof of concept that use a functionnal approach.

	\paragraph*{}
        Used technologies :
        \begin{description}
                \item[python] as the main language (version 3.4);
                \item[docopt] as command-line arguments parser;
                \item[doctest] for unit tests in some docstrings (see \textsc{unittests.py} file);
                \item[numpy] for manage matrix (and notabily multiplications) until Python 3.5 interpreter is available;
                \item[latex] for present report, with pgfplot module for automatic data vizualization generation;
                \item[pylint] for source code profiling and UML generation;
                \item[make] for automatize simulations, tests, etc;
        \end{description}





%%%%%%%%%%%%%%%%%%%%%%%%%
% SECTION               %
%%%%%%%%%%%%%%%%%%%%%%%%%
\section*{Source code architecture}
	\paragraph*{}
        Genomat is cut in modules as shown in Figure~\ref{fig:umldiag}, and described below.
        \begin{description}
                \item[main] implements a simple and efficient interface that allow command-line arguments;
                \item[config] management of configurations : files, options, flags, serialization;
                \item[gene network] definition of an individual, encapsulation of numpy matrix system;
                \item[population] a list of individuals that provides many data access and call stats modules as an observer;
                \item[stats] observer of population that tests it and generate, if asked, data in targeted files;
                \item[progress bar] dumb and simple implementation of a progress bar, designed to be replaced later;
        \end{description}

        \begin{figure}[H] 
                \centering
                \includegraphics[width=\textwidth]{packages_genomat.png}
                \caption{Modules of Genomat architecture, without externals ones (numpy, docopt).}
                \label{fig:umldiag}
        \end{figure}




%%%%%%%%%%%%%%%%%%%%%%%%%
% SECTION               %
%%%%%%%%%%%%%%%%%%%%%%%%%
\section{Watchables}
    \paragraph*{}
    Main data generated results of gene KO, for each gene at each generation. 
    A diversity ratio is also computed at each generation.
    \paragraph*{}
     First, as there was no mutation rate given, exploited simulations use a mutation rate from $10^{-0}$ to $10^{-6}$, 
     with a constant population size of 300 individuals and 200 generations.\\
     For each gene, the survavibility ratio is computed. \\The survivability ratio is defined as $S_{g} = \frac{KO_g}{population_{size}}$ 
     with $KO_g$ as the number of individuals still viable after the KO of gene g.
    \paragraph*{}
     With those parameters, we can notice the five genes have a similar course : their survivability ratio first fastly 
     increase from 0.7 during the fifty first generations. After this exponentianal phase, the survavibility ratio reach a plateau about 1. 
    \paragraph*{}





%%%%%%%%%%%%%%%%%%%%%%%%%
% SECTION               %
%%%%%%%%%%%%%%%%%%%%%%%%%
\section*{Results through graphics}
    \paragraph*{}
    The used telltale follow the next rules :
    \simulationParameters{}





%%%%%%%%%%%%%%%%%%%%%%%%%
% SECTION               %
%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%%%%%%%%
% APPENDICES            %
%%%%%%%%%%%%%%%%%%%%%%%%%
%\newpage
%\begin{appendix}
%\section*{Annexes}

%\begin{figure}[h] % place it THERE (add ! for exact placement)
        %\centering
        %\pgfplotstabletypeset[
            %columns={time,wtf,error},
            %col sep=comma,
            %string type,
            %every head row/.style={%
                %before row={\hline
                    %& \multicolumn{2}{c}{useless data} \\
                    %},
                %after row=\hline
            %},
            %every last row/.style={after row=\hline},
            %columns/time/.style={column name=Time, column type=l},
            %columns/error/.style={column name=Primary key, column type=r},
            %columns/wtf/.style={column name=Concentration, column type=r},
        %]{data.csv}
        %\caption{Cool things that will be very useful for someone}
        %\label{ftab:cool}
%\end{figure}      



%\end{appendix}
\end{document}
% END

